### ISSUE TITLE
Proposal to improve search

### ISSUE THREAD

**Issue by glow12121** [appml/neutrinote#64]
**Is your feature request related to a problem? Please describe.**
There is search functionality in the general list of notes. It is, however, not possible to jump between the highlighted search results.

Also, it isn't possible to search when in the note-editing mode.

**Describe the solution you'd like**
It would be nice to improve search functionality.

**Comment by appml** [comment]
To conduct search during edit mode, simply tap search icon in the tool bar.  To navigate, tap the search icon again or swipe up / down at the status region as shown:


![Screenshot_20230101-095559](https://user-images.githubusercontent.com/7504229/210175069-7aa2e73c-c8e0-4788-ad6a-64f579b7e181.png)

**Comment by glow12121** [comment]
Thank you. But:

1. Both the fact that the search icon is hidden in tools which itself can even be hidden -- and the up/down jumps are most counterintuitive.
2. The following error occurs on tap on the upper thing. The copied fragment in buffer is copied, plus a cryptic message..

![Screenshot_20230101-204345_neutriNote](https://user-images.githubusercontent.com/29035180/210176860-b7a84737-a5c9-4609-a878-850e09a6e707.png)

**Comment by appml** [comment]
Not sure what you mean by 1.  

Regarding 2, it is just one of the ways to navigate search results.  Please see [Gesture](https://appml.github.io/nano/#gestures) section.

**Comment by glow12121** [comment]
It's hard to find the search icon and after you find it, there's no up/down buttons.

**Comment by appml** [comment]
> It's hard to find the search icon and after you find it, there's no up/down buttons.
> 
> The message ("W: ... S: ...") doesn't seem to make sense.

Sorry I am not sure what you mean by hard to find.  Which ROM / Android version do you have?

Also there are more than one way to navigate the search results.

![Screenshot_20230103-183027](https://user-images.githubusercontent.com/7504229/210458707-bdb40c0c-3fec-4152-ab0c-2085b06bf0ef.png)

**Comment by glow12121** [comment]
What I mean is just - how the search icon is located is counter-intuitive (especially given where search is in the notes list) and hard to find. I don't know how else to explain.

Same with absense of up/down arrows in the search box. Additional gestures may be good but little arrows is what everyone is used to.

**Comment by GfEW** [comment]
@glow12121 
> It's hard to find the search icon and after you find it, there's no up/down buttons.
> 
> The message ("W: ... S: ...") doesn't seem to make sense.

I'm new to Neutrinote CE and in the process of deciding whether it makes sense for me to start using it for real stuff, therefore browsing through documentation and issues. 

The "cryptic" message you're referring to appears to give the count of words and the document size, respectively. In long notes, there may also be a third number with label `L:` that denotes the line number of the currently focussed match.

Although all those numbers can be useful, I agree the design, place, and way to get this little message displayed, are a bit arcane and not self explanatory. To say the least, the one letter abbreviations should be explained somewhere nearby. Also, those numbers do not make up for the lack of proper ___previous___ and ___next___ buttons.

The search results view could be much more complete if it allowed the user to quickly jump to the previous or next search result with a single tap. Currently, tapping the search icon again, as well as F3 on the keyboard, do work as _next_, but there doesn't seem to be any means to go _previous_ (unless it's even more hidden so I've not discovered it yet). To freely navigate search results back and forth, at least Shift+F3 should jump to the _previous_ result.

Additionally, a __compact lilst view of all matching lines__ that can be jumped to directly would also be very useful.

**Comment by glow12121** [comment]
> I'm new to Neutrinote CE and, in the process of deciding whether it makes sense for me to start using it for real notes

I have been using Neutrinote for several years now and am not planning to switch. The main reason is that it works well with large files.

Many thanks to the developer but, yes, these interface features are counter-intuitive.

**Comment by GfEW** [comment]
> By the way, what is C.E.?

I don't know for sure, but _CE_ might stand for [Community Edition](https://www.f-droid.org/en/packages/com.appmindlab.nano/).

Thanks for sharing your experience, good to know NN has proven worthwhile for you over the years. I'm still deterred by a few crucial issues, with this one being only the fifth:
1. [make regexes fully functional in global search](#76)
2. auto-bullet/auto-indent to continue current bullet style / indentation depth upon entering a newline
3. buttons or drop-down suggestions for global search field operators (`title`, `titlereg:` etc.)
4. make the entire app keyboard-accessible 
5. [this very issue](#64) on improving the in-file search UX

I guess I'll try and file them one by one, to see if there is any resonance on improving them.

**Comment by GfEW** [comment]
@glow12121

BTW, I think <details><summary>the red arrows in this screenshot</summary> ![Screenshot_20230101-095559](https://user-images.githubusercontent.com/7504229/210175069-7aa2e73c-c8e0-4788-ad6a-64f579b7e181.png)
</details>
are not meant by @appml to point to the match count or search icon, specifically, but simply to indicate "swipe up / swipe down" scrolling directions. 

Of course, scrolling cannot make up for the lack of __previous__ or __next__ buttons, either.

**Comment by glow12121** [comment]
> the red arrows in this screenshot

But when you click the upper arrow area, I think, the notepad pastes the content copied in the buffer...

**Comment by GfEW** [comment]
> > the red arrows in this screenshot
> 
> But when you click the upper arrow area, I think, the notepad pastes the content copied in the buffer...

I can't seem to reproduce that. The area doesn't show any special behaviour here, it just puts the text cursor at the place of my tap.

**Comment by glow12121** [comment]
@GfEW I've tried this again. You are right, the copied fragment is not pasted in the text area but, in my case, rather pasted in this upper right area pointed by the red arrow.

Say, I have "aaa" in buffer.

I search "s".

Tapping on the upper right area triggers a) the box with the characters count b) unexpectedly, the buffer fragment pasted in the upper right area itself.

**Comment by glow12121** [comment]
@GfEW Generally speaking, it would perhaps make sense to make the two search interfaces more consistent. By the first one I mean searching from the notes list (there's a logical magnifying glass icon in the top bar). By the second - searching from within a note.

**Comment by appml** [comment]
@GfEW I agree that the UI is not exposing some of the functionalities.  The app was originally designed almost 10 years ago and has accumulated more than 1000 builds so far, so the priority now is to retain as much as features across iterations of Android versions (several already been dropped over the years).  As you might have noticed if you follow the app's development a while, most recent new features are CLI oriented.  

Ideally someone could come along and revamped the UI frontend, or carry on with all the features with something better, or fork it into an entirely new app, as the current development resource is tightly limited.

**Comment by GfEW** [comment]
@appml 
I greatly appreciate your explanation, it's perfectly understandable. Since I'm not a coder, my contribution to facilitating more active development most likely won't be on the code side of things. However, if I find time to elaborate concise, reasonable, comprehensible improvement suggestions, then whoever may want to pick up NN development already finds some ideas to build on.

**Comment by GfEW** [comment]
> (b) [...] displaying, of all things, the clipboard contents there [...] appears to be a bug, or a _very_ weird UI design decision.
On (a), we already agree.

As it turns out, that's really by design, and there is a previous/next navigation in search results already! It's all explained [here](https://appml.github.io/nano/#gestures).

What remains is the obvious non-obviousness of these features, in terms of a modern, self-explanatory GUI. Which only reinforces the [need for a revamp](https://github.com/appml/neutrinote/issues/64#issuecomment-2094675493). 

I think that Neutrinote has invaluable potential to be a fast, focussed, lightweight alternative to Obsidian, Joplin and the likes, but as of now, there are too many rough edges desparately looking for someone to polish them, if not to give the GUI a general, systematic overhaul.
